sympy/solvers/tests/test_solveset.py

```diff
@@ -2323,7 +2233,111 @@ def test_solve_logarithm():
     y = Symbol('y', positive=True)
     assert _solve_logarithm(log(x)*log(y), 0, x, S.Reals) == FiniteSet(1)
 
-# end of logarithmic tests
+#  end of logarithmic tests
+
+
+# lambert tests for solveset
+def test_solve_lambert():
+    assert solveset_real(x*exp(x) - 1, x) == FiniteSet(LambertW(1))
+    assert solveset_real(exp(x) + x, x) == FiniteSet(-LambertW(1))
+    assert solveset_real(x + 2**x, x) == \
+        FiniteSet(-LambertW(log(2))/log(2))
+
+    assert solveset_real(3*x + log(4*x), x) == \
+        FiniteSet(LambertW(Rational(3, 4))/3)
+
+    a = Symbol('a', real=True)
+    assert solveset_real(a/x + exp(x/2), x) == \
+        Complement(FiniteSet(2*LambertW(-a/2)), FiniteSet(0))
+    assert solveset_real((a/x + exp(x/2)).diff(x), x) == \
+        Complement(FiniteSet(4*LambertW(-sqrt(2)*sqrt(a)/4),\
+             4*LambertW(sqrt(2)*sqrt(a)/4)), FiniteSet(0))
+
+    # coverage test
+    assert solveset_real(tanh(x + 3)*tanh(x - 3) - 1, x) == EmptySet()
+
+    assert solveset_real(x*log(x) + 3*x + 1, x) == \
+        FiniteSet(exp(-3 + LambertW(-exp(3))))
+    eq = (x*exp(x) - 3).subs(x, x*exp(x))
+    assert solveset_real(eq, x) == \
+        FiniteSet(LambertW(3*exp(-LambertW(3))))
+
+    assert solveset_real(3*log(a**(3*x + 5)) + a**(3*x + 5), x) == \
+        FiniteSet((-log(a**5) - LambertW(Rational(1, 3)))/(3*log(a)))
+
+    # check collection
+    b = Symbol('b')
+    eq = 3*log(a**(3*x + 5)) + b*log(a**(3*x + 5)) + a**(3*x + 5)
+    assert solveset_real(eq, x) == FiniteSet((-log(a**5) - LambertW(1/(b + 3)))/(3*log(a)))
+
+
+    assert solveset_real(x**3 - 3**x, x) == \
+        FiniteSet(-3*LambertW(-log(3)/3)/log(3), -3*LambertW(-log(3)/3, -1)/log(3))
+
+    assert solveset_real(x**2 - 2**x, x) == \
+        solveset_real(-x**2 + 2**x, x)
+
+    assert solveset_real(3*log(x) - x*log(3), x) == FiniteSet(
+        -3*LambertW(-log(3)/3)/log(3),
+        -3*LambertW(-log(3)/3, -1)/log(3))
+
+    assert solveset_real(LambertW(2*x) - y, x) == \
+        Intersection(FiniteSet(y*exp(y)/2), S.Reals)
```

Comment:
This test is good, but I think it would also be good to have one involving a real symbol `yr = Symbol('yr', real=True)`.